sia-antfarm/antfarm.go

```diff
@@ -0,0 +1,170 @@
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/julienschmidt/httprouter"
+)
+
+type (
+	// AntfarmConfig contains the fields to parse and use to create a sia-antfarm.
+	AntfarmConfig struct {
+		ListenAddress string
+		AntConfigs    []AntConfig
+		AutoConnect   bool
+
+		// ExternalFarms is a slice of net addresses representing the API addresses
+		// of other ant farms to connect to.
+		ExternalFarms []string
+	}
+
+	// antFarm defines the 'antfarm' type. antFarm orchestrates a collection of
+	// ants and provides an API server to interact with them.
+	antFarm struct {
+		apiListener net.Listener
+		wg          sync.WaitGroup
+		ants        []*Ant
+		router      *httprouter.Router
+	}
+)
+
+// createAntfarm creates a new antFarm given the supplied AntfarmConfig
+func createAntfarm(config AntfarmConfig) (*antFarm, error) {
+	// clear old antfarm data before creating an antfarm
+	os.RemoveAll("./antfarm-data")
+	os.MkdirAll("./antfarm-data", 0700)
+
+	farm := &antFarm{}
+
+	// start up each ant process with its jobs
+	ants, err := startAnts(config.AntConfigs...)
+	if err != nil {
+		return nil, err
+	}
+	farm.ants = ants
+	farm.wg.Add(len(ants))
+	go func() {
+		for _, ant := range ants {
+			ant.Wait()
+			farm.wg.Done()
+		}
+	}()
+
+	err = func() error {
+		// if the AutoConnect flag is set, use connectAnts to bootstrap the network.
+		if config.AutoConnect {
+			if err = connectAnts(ants...); err != nil {
+				return err
+			}
+		}
+		// connect the external ant farms
+		for _, address := range config.ExternalFarms {
+			if err = farm.connectExternalAntfarm(address); err != nil {
+				return err
+			}
+		}
+		// start up the api server listener
+		farm.apiListener, err = net.Listen("tcp", config.ListenAddress)
+		if err != nil {
+			return err
+		}
+		return nil
+	}()
+
+	if err != nil {
```

Comment:
error check should not be split from the logic that creates the error (remove newline)